### PR TITLE
[io] Improve PCD read performance (more than 50%) by reusing `istringstream`

### PR DESCRIPTION
**Pull request by kunaltyagi**
Single threaded performance increases by approximate 50%

https://github.com/PointCloudLibrary/pcl/issues/4333#issuecomment-676583821

### CODE REVIEW

File: io/include/pcl/io/file_io.h [hunk 1]
```diff
@@ -324,6 +324,83 @@ namespace pcl
...
+    copyStringValue<std::uint8_t> (const std::string &st, pcl::PCLPointCloud2 &cloud,
+                            pcl::index_t point_index, unsigned int field_idx, unsigned int fields_count,
+                    std::istringstream& is)
+    {
+      std::uint8_t value;
+      if (boost::iequals(st, "nan"))
+      {
+        value = static_cast<std::uint8_t> (std::numeric_limits<int>::quiet_NaN ());
+        cloud.is_dense = false;
+      }
```
Comment (SergioRAgostinho) [inline io/include/pcl/io/file_io.h:387]: This should never happen, an integer type doesn't have a nan representation.

Comment (kunaltyagi) [reply io/include/pcl/io/file_io.h:387]: Yeah. Replace with 0 (the actual value of the expression) or remove the condition (change behavior)?

Comment (SergioRAgostinho) [reply io/include/pcl/io/file_io.h:387]: I would say remove the condition altogether.

**Comment by SergioRAgostinho** [comment]
I'm not going to review this one. I have no progress on it.

**Comment by mvieth** [comment]
The changes look ok, but I feel that there has to be a nicer way to do all this. What about `stod`, `stof`, `stoi`, ...? Alternatively, maybe the `boost:split` in pcd_io.cpp can be removed and instead the stream used for the whole line?

**Comment by kunaltyagi** [comment]
> What about `stod`, `stof`, `stoi`, ...?

AFAIK, these functions are locale dependent. In C, the locale is global and changing locale might affect downstream users during loading of a file.

> Alternatively, maybe the `boost:split` in pcd_io.cpp can be removed and instead the stream used for the whole line?

I've marked that issue as help-wanted because I couldn't refactor it cleanly without touching quite a few lines. I too would like to see a cleaner implementation than this low hanging fruit.

**Review by larshg — APPROVED** [review]
I agree to @mvieth observations, but as you write @kunaltyagi it can be left for another day :)

50% performance increase, is also quite nice!

File: io/include/pcl/io/file_io.h [hunk 2]
```diff
@@ -324,90 +324,107 @@ namespace pcl
...
+  template <typename Type> inline void
+  copyStringValue (const std::string &st, pcl::PCLPointCloud2 &cloud,
+                   pcl::index_t point_index, unsigned int field_idx, unsigned int fields_count)
+  {
+    std::istringstream is;
+    is.imbue (std::locale::classic ());
+    detail::copyStringValue<Type> (st, cloud,point_index, field_idx, fields_count, is);
+  }
+  template <typename Type> inline void
+  copyStringValue (const std::string &st, pcl::PCLPointCloud2 &cloud,
```
Comment (mvieth) [inline io/include/pcl/io/file_io.h:424]: This new functions lacks the documentation, especially regarding `is`. Also: "Uses aoti/atof to do the conversion." isn't really correct anymore.

**Review by mvieth — APPROVED** [review]

**Pull request merged**